pkg/network/http/http_statkeeper.go

```diff
@@ -4,51 +4,63 @@ package http
 
 import (
 	"C"
-	"sync"
 )
 
 type httpStatKeeper struct {
-	mux   sync.Mutex
-	stats map[Key]map[string]RequestStats
+	stats map[Key]RequestStats
+
+	// http path buffer
+	buffer []byte
+
+	// map containing interned path strings
+	// this is rotated  with the stats map
+	interned map[string]string
 }
 
 func newHTTPStatkeeper() *httpStatKeeper {
 	return &httpStatKeeper{
-		stats: make(map[Key]map[string]RequestStats),
+		stats:    make(map[Key]RequestStats),
+		buffer:   make([]byte, HTTPBufferSize),
+		interned: make(map[string]string),
 	}
 }
 
 func (h *httpStatKeeper) Process(transactions []httpTX) {
-	h.mux.Lock()
-	defer h.mux.Unlock()
-
 	for _, tx := range transactions {
-		key := Key{
-			SourceIP:   tx.SourceIP(),
-			DestIP:     tx.DestIP(),
-			SourcePort: tx.SourcePort(),
-			DestPort:   tx.DestPort(),
-		}
-		path := tx.Path()
-		statusClass := tx.StatusClass()
-		latency := tx.RequestLatency()
-
-		if _, ok := h.stats[key]; !ok {
-			h.stats[key] = make(map[string]RequestStats)
-		}
-		stats := h.stats[key][path]
-		stats.AddRequest(statusClass, latency)
-		h.stats[key][path] = stats
+		key := h.newKey(tx)
+		stats := h.stats[key]
+		stats.AddRequest(tx.StatusClass(), tx.RequestLatency())
+		h.stats[key] = stats
 	}
 }
 
-func (h *httpStatKeeper) GetAndResetAllStats() map[Key]map[string]RequestStats {
-	h.mux.Lock()
-	defer h.mux.Unlock()
-
+func (h *httpStatKeeper) GetAndResetAllStats() map[Key]RequestStats {
 	ret := h.stats // No deep copy needed since `h.stats` gets reset
-	h.stats = make(map[Key]map[string]RequestStats)
+	h.stats = make(map[Key]RequestStats)
+	h.interned = make(map[string]string)
 	return ret
 }
 
-func (h *httpStatKeeper) Close() {}
+func (h *httpStatKeeper) newKey(tx httpTX) Key {
+	path := tx.Path(h.buffer)
+	pathString := h.intern(path)
+
+	return Key{
+		SrcIPHigh: uint64(tx.tup.saddr_h),
+		SrcIPLow:  uint64(tx.tup.saddr_l),
+		SrcPort:   uint16(tx.tup.sport),
+		DstIPHigh: uint64(tx.tup.daddr_h),
+		DstIPLow:  uint64(tx.tup.daddr_l),
+		DstPort:   uint16(tx.tup.dport),
+		Path:      pathString,
+	}
+}
+
+func (h *httpStatKeeper) intern(b []byte) string {
+	v, ok := h.interned[string(b)]
```

Comment:
I know there were some optimizations to have this not allocate in recent versions of go -- does this string conversion wind up allocating? Since `b` is usually immutable it would be nice to do the map lookup without allocating, and only allocate when we need to intern a new string.

Reply:
Yup, there is an optimization in Golang for this type of lookup: https://github.com/golang/go/commit/f5f5a8b6209f84961687d993b93ea0d397f5d5bf